File: mobile/src/main/java/net/mm2d/dmsexplorer/viewmodel/helper/MovieActivityPipHelperOreo.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.viewmodel.helper;

import android.app.Activity;
import android.app.PendingIntent;
import android.app.PictureInPictureParams;
import android.app.RemoteAction;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Rect;
import android.graphics.drawable.Icon;
import android.os.Build;
import android.text.TextUtils;
import android.util.Rational;
import android.view.View;

import net.mm2d.dmsexplorer.Const;
import net.mm2d.dmsexplorer.R;
import net.mm2d.dmsexplorer.viewmodel.ControlPanelModel;
import net.mm2d.log.Log;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.annotation.StringRes;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
@RequiresApi(api = Build.VERSION_CODES.O)
class MovieActivityPipHelperOreo implements MovieActivityPipHelper {
    @NonNull
    private final Activity mActivity;
    @Nullable
    private ControlPanelModel mControlPanelModel;
    @NonNull
    private final BroadcastReceiver mControlReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(
                Context context,
                Intent intent) {
            final String action = intent.getAction();
            if (TextUtils.isEmpty(action) || mControlPanelModel == null) {
                return;
            }
            switch (intent.getAction()) {
                case Const.ACTION_PLAY:
                    mControlPanelModel.onClickPlayPause();
                    final PictureInPictureParams.Builder builder = new PictureInPictureParams.Builder();
                    builder.setActions(makeActions(mControlPanelModel.isPlaying()));
                    mActivity.setPictureInPictureParams(builder.build());
                    break;
                case Const.ACTION_NEXT:
                    mControlPanelModel.onClickNext();
                    break;
                case Const.ACTION_PREV:
                    mControlPanelModel.onClickPrevious();
                    break;
            }
        }
    };

    MovieActivityPipHelperOreo(@NonNull final Activity activity) {
        mActivity = activity;
    }

    @Override
    public void register() {
        mActivity.registerReceiver(mControlReceiver, makeIntentFilter());
    }

    @NonNull
    private IntentFilter makeIntentFilter() {
        final IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(Const.ACTION_PLAY);
        intentFilter.addAction(Const.ACTION_NEXT);
        intentFilter.addAction(Const.ACTION_PREV);
        return intentFilter;
    }

    @Override
    public void unregister() {
        mActivity.unregisterReceiver(mControlReceiver);
    }

    @Override
    public void setControlPanelModel(@Nullable final ControlPanelModel model) {
        mControlPanelModel = model;
    }

    @Override
    public void enterPictureInPictureMode(@NonNull final View contentView) {
        final PictureInPictureParams.Builder builder = new PictureInPictureParams.Builder();
        final List<RemoteAction> actions = makeActions(mControlPanelModel.isPlaying());
        if (!actions.isEmpty()) {
            builder.setActions(actions);
        }
        final Rect rect = makeViewRect(contentView);
        if (rect.width() > 0 && rect.height() > 0) {
            builder.setSourceRectHint(rect)
                    .setAspectRatio(new Rational(rect.width(), rect.height()));
        }
        try {
            mActivity.enterPictureInPictureMode(builder.build());
        } catch (final Exception e) {
            Log.w(e);
        }
    }

    @NonNull
    private Rect makeViewRect(@NonNull final View v) {
        final Rect rect = new Rect();
        v.getGlobalVisibleRect(rect);
        return rect;
    }

    @NonNull
    private List<RemoteAction> makeActions(final boolean isPlaying) {
        final int max = mActivity.getMaxNumPictureInPictureActions();
        if (max <= 0) {
            return Collections.emptyList();
        }
        if (max >= 3) {
            return Arrays.asList(
                    makePreviousAction(),
                    makePlayAction(isPlaying),
                    makeNextAction()
            );
        }
        return Collections.singletonList(makePlayAction(isPlaying));
    }

    @NonNull
    private RemoteAction makePlayAction(final boolean isPlaying) {
        return new RemoteAction(
                makeIcon(isPlaying ? R.drawable.ic_pause : R.drawable.ic_play),
                getString(R.string.action_play_title),
                getString(R.string.action_play_description),
                makePlayPendingIntent(mActivity));
    }

    @NonNull
    private RemoteAction makeNextAction() {
        return new RemoteAction(
                makeIcon(R.drawable.ic_skip_next),
                getString(R.string.action_next_title),
                getString(R.string.action_next_description),
                makeNextPendingIntent(mActivity));
    }

    @NonNull
    private RemoteAction makePreviousAction() {
        return new RemoteAction(
                makeIcon(R.drawable.ic_skip_previous),
                getString(R.string.action_previous_title),
                getString(R.string.action_previous_description),
                makePreviousPendingIntent(mActivity));
    }

    @NonNull
    private String getString(@StringRes final int resId) {
        return mActivity.getResources().getText(resId, "").toString();
    }

    private Icon makeIcon(@DrawableRes final int resId) {
        return Icon.createWithResource(mActivity, resId);
    }

    @NonNull
    private static PendingIntent makePlayPendingIntent(@NonNull final Context context) {
        return PendingIntent.getBroadcast(context,
                Const.REQUEST_CODE_ACTION_PLAY,
                new Intent(Const.ACTION_PLAY),
                PendingIntent.FLAG_UPDATE_CURRENT);
    }

    @NonNull
    private static PendingIntent makeNextPendingIntent(@NonNull final Context context) {
        return PendingIntent.getBroadcast(context,
                Const.REQUEST_CODE_ACTION_NEXT,
                new Intent(Const.ACTION_NEXT),
                PendingIntent.FLAG_UPDATE_CURRENT);
    }

    @NonNull
    private static PendingIntent makePreviousPendingIntent(@NonNull final Context context) {
        return PendingIntent.getBroadcast(context,
                Const.REQUEST_CODE_ACTION_PREVIOUS,
                new Intent(Const.ACTION_PREV),
                PendingIntent.FLAG_UPDATE_CURRENT);
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/viewmodel/helper/MovieActivityPipHelperOreo.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.viewmodel.helper

import android.app.Activity
import android.app.AppOpsManager
import android.app.PendingIntent
import android.app.PictureInPictureParams
import android.app.RemoteAction
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Rect
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Build
import android.util.Rational
import android.view.View
import androidx.annotation.DrawableRes
import androidx.annotation.RequiresApi
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import net.mm2d.dmsexplorer.Const
import net.mm2d.dmsexplorer.R
import net.mm2d.dmsexplorer.viewmodel.ControlPanelModel
import net.mm2d.log.Logger

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
@RequiresApi(api = Build.VERSION_CODES.O)
internal class MovieActivityPipHelperOreo(
    private val activity: Activity,
) : MovieActivityPipHelper {
    private var controlPanelModel: ControlPanelModel? = null
    private val controlReceiver = object : BroadcastReceiver() {
        override fun onReceive(
            context: Context,
            intent: Intent,
        ) {
            val action = intent.action
            val model = controlPanelModel
            if (action.isNullOrEmpty() || model == null) {
                return
            }
            when (action) {
                Const.ACTION_PLAY -> {
                    model.onClickPlayPause()
                    activity.setPictureInPictureParams(
                        PictureInPictureParams.Builder()
                            .setActions(makeActions(model.isPlaying))
                            .build(),
                    )
                }
                Const.ACTION_NEXT -> model.onClickNext()
                Const.ACTION_PREV -> model.onClickPrevious()
            }
        }
    }

    private fun isPictureInPictureAllowed(): Boolean {
        val appOps = activity.getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
        return try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                appOps.unsafeCheckOpNoThrow(
                    AppOpsManager.OPSTR_PICTURE_IN_PICTURE,
                    android.os.Process.myUid(),
                    activity.packageName,
                )
            } else {
                @Suppress("DEPRECATION")
                appOps.checkOpNoThrow(
                    AppOpsManager.OPSTR_PICTURE_IN_PICTURE,
                    android.os.Process.myUid(),
                    activity.packageName,
                )
            } == AppOpsManager.MODE_ALLOWED
        } catch (ignored: Exception) {
            false
        }
    }

    override fun register() {
        ContextCompat.registerReceiver(
            activity,
            controlReceiver,
            makeIntentFilter(),
            ContextCompat.RECEIVER_EXPORTED,
        )
    }

    private fun makeIntentFilter(): IntentFilter = IntentFilter().also {
        it.addAction(Const.ACTION_PLAY)
        it.addAction(Const.ACTION_NEXT)
        it.addAction(Const.ACTION_PREV)
    }

    override fun unregister() {
        activity.unregisterReceiver(controlReceiver)
    }

    override fun setControlPanelModel(model: ControlPanelModel?) {
        controlPanelModel = model
    }

    override fun enterPictureInPictureMode(contentView: View) {
        if (!isPictureInPictureAllowed()) {
            val intent = Intent(ACTION_PICTURE_IN_PICTURE_SETTINGS)
            intent.data = Uri.parse("package:" + activity.packageName)
            activity.startActivity(intent)
            return
        }
        val builder = PictureInPictureParams.Builder()
        builder.setActions(controlPanelModel?.isPlaying?.let { makeActions(it) })
        val rect = makeViewRect(contentView)
        if (rect.width() > 0 && rect.height() > 0) {
            builder.setSourceRectHint(rect)
                .setAspectRatio(Rational(rect.width(), rect.height()))
        }
        try {
            activity.enterPictureInPictureMode(builder.build())
        } catch (e: Exception) {
            Logger.w(e)
        }
    }

    private fun makeViewRect(v: View): Rect = Rect().also {
        v.getGlobalVisibleRect(it)
    }

    private fun makeActions(isPlaying: Boolean): List<RemoteAction>? {
        val max = activity.maxNumPictureInPictureActions
        if (max <= 0) {
            return null
        }
        return if (max >= 3) {
            listOf(makePreviousAction(), makePlayAction(isPlaying), makeNextAction())
        } else {
            listOf(makePlayAction(isPlaying))
        }
    }

    private fun makePlayAction(isPlaying: Boolean): RemoteAction = RemoteAction(
        makeIcon(if (isPlaying) R.drawable.ic_pause else R.drawable.ic_play),
        getString(R.string.action_play_title),
        getString(R.string.action_play_description),
        makePlayPendingIntent(activity),
    )

    private fun makeNextAction(): RemoteAction = RemoteAction(
        makeIcon(R.drawable.ic_skip_next),
        getString(R.string.action_next_title),
        getString(R.string.action_next_description),
        makeNextPendingIntent(activity),
    )

    private fun makePreviousAction(): RemoteAction = RemoteAction(
        makeIcon(R.drawable.ic_skip_previous),
        getString(R.string.action_previous_title),
        getString(R.string.action_previous_description),
        makePreviousPendingIntent(activity),
    )

    private fun getString(@StringRes resId: Int): String =
        activity.resources.getText(resId, "").toString()

    private fun makeIcon(@DrawableRes resId: Int): Icon = Icon.createWithResource(activity, resId)

    companion object {
        private const val ACTION_PICTURE_IN_PICTURE_SETTINGS =
            "android.settings.PICTURE_IN_PICTURE_SETTINGS"

        private fun makePlayPendingIntent(context: Context): PendingIntent =
            PendingIntent.getBroadcast(
                context,
                Const.REQUEST_CODE_ACTION_PLAY,
                Intent(Const.ACTION_PLAY),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )

        private fun makeNextPendingIntent(context: Context): PendingIntent =
            PendingIntent.getBroadcast(
                context,
                Const.REQUEST_CODE_ACTION_NEXT,
                Intent(Const.ACTION_NEXT),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )

        private fun makePreviousPendingIntent(context: Context): PendingIntent =
            PendingIntent.getBroadcast(
                context,
                Const.REQUEST_CODE_ACTION_PREVIOUS,
                Intent(Const.ACTION_PREV),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )
    }
}
